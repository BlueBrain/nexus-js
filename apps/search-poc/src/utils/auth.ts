import { Realm } from '@bbp/nexus-sdk';
import { UserManager, User } from 'oidc-client';
import { Operation, Link, Observable } from '@bbp/nexus-link';
import { SETTINGS } from '../config';

// creates a OIDC config based on input realm
export function getConfig(realm?: Realm) {
  return {
    authority: realm ? realm._issuer : '',
    client_id: SETTINGS.clientId,
    redirect_uri: SETTINGS.redirectUrl,
    popup_redirect_uri: SETTINGS.redirectUrl,
    post_logout_redirect_uri: SETTINGS.redirectUrl,
    response_type: 'id_token token',
    loadUserInfo: true,
  };
}

// link that sets the bearer token before each nexus call
export const setToken: Link = (operation: Operation, forward?: Link) => {
  const token = localStorage.getItem(SETTINGS.bearerTokenKey);
  const nextHeaders: any = { ...operation.headers };
  token && (nextHeaders['Authorization'] = `Bearer ${token}`);

  const nextOperation = {
    ...operation,
    headers: nextHeaders,
  };

  return forward ? forward(nextOperation) : new Observable();
};

/**
 * This function will:
 * - try to create a user manager instance if the realms settings are available from local storage
 * - set event listeners in order to keep the bearer token in sync in local storage
 * - look for new session on creation
 * - load user info
 * - return a ref of user manager instance and the user
 */
export async function setUpSession(): Promise<[UserManager, User | null]> {
  let user: User | null = null;
  let userManager: UserManager | null = null;
  // Get preferred realm
  const preferredRealm: Realm = JSON.parse(
    localStorage.getItem(SETTINGS.preferredRealmKey) || '{}',
  );

  // then create User Manager with preferred realm issuer
  // or empty user manager if no preferred realm
  userManager = new UserManager(getConfig(preferredRealm));

  // maybe we're getting new user session after redirect
  try {
    user = await userManager.signinRedirectCallback();
  } catch (error) {}
  // if not, check if we already had a running session
  if (!user) {
    try {
      user = await userManager.getUser();
    } catch (error) {
      // if not, that's fine
    }
  }

  // Set events
  userManager.events.addUserLoaded((user: User) => {
    localStorage.setItem(SETTINGS.bearerTokenKey, user.id_token);
  });

  userManager.events.addAccessTokenExpired(() => {
    localStorage.removeItem(SETTINGS.bearerTokenKey);
  });

  return [userManager, user];
}
