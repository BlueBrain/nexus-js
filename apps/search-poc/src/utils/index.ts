/**
 * Gets the last part of a uri path which represent a title or label
 *
 * @param string
 */
export const getLabel = (string: string): string => {
  const slash = string.substring(string.lastIndexOf('/') + 1);
  const title = slash.substring(slash.lastIndexOf('#') + 1);
  return title;
};

export const parseProjectUrl = (projectUrl: string) => {
  const projectUrlR = /projects\/([\w-]+)\/([\w-]+)\/?$/;
  const [, org, proj] = projectUrl.match(projectUrlR) as string[];
  return [org, proj];
};

export type ParsedNexusUrl = {
  deployment: string;
  entityType: string;
  org: string;
  project: string;
  schema: string;
  id: string;
};

const nexusEntities = [
  'orgs',
  'projects',
  'acls',
  'views',
  'resources',
  'files',
];

const nexusUrlR = new RegExp(
  [
    '^',
    '(https?://.+)', // nexus deployment
    '/',
    `(${nexusEntities.join('|')})`, // entity type
    '/',
    '([^/]+)', // org
    '/',
    '([^/]+)', // proj
    '/?',
    '([^/]+)?', // schema [optional]
    '/?',
    '([^/]+)?', // id [optional]
    '/?',
    '$',
  ].join(''),
);

/**
 * With given Nexus URL (might be self/project/id url), return it's:
 * * deployment URL
 * * entity type
 * * org label
 * * project label
 * * id
 *
 * @param nexusUrl
 */
export const parseNexusUrl = (nexusUrl: string): ParsedNexusUrl => {
  if (!nexusUrl) throw new Error('selfUrl should be defined');

  const mulEntityTypeR = new RegExp(`(${nexusEntities.join('|')})`, 'g');
  const mulEntityTypeMatch = nexusUrl.match(mulEntityTypeR);
  if (mulEntityTypeMatch && mulEntityTypeMatch.length > 1) {
    throw new Error(
      'Url contains multiple entity types which is not supported',
    );
  }

  const matches = nexusUrl.match(nexusUrlR);
  if (!matches || matches.length <= 5) {
    throw new Error('Error while parsing selfUrl');
  }

  return {
    deployment: matches[1],
    entityType: matches[2].slice(0, -1),
    org: matches[3],
    project: matches[4],
    schema: matches[5],
    id: matches[6],
  };
};

/**
 * this function changes cameCasedString to Camel Cased String
 * @author https://stackoverflow.com/questions/4149276/how-to-convert-camelcase-to-camel-case
 * @param labelString String in camelCase
 */
export const camelCaseToLabelString = (labelString: string): string => {
  return (
    labelString
      // insert a space before all caps
      .replace(/([A-Z])/g, ' $1')
      // uppercase the first character
      .replace(/^./, str => str.toUpperCase())
  );
};

/**
 * this function changes camelCaseString to KebabCase
 * @author https://gist.github.com/nblackburn/875e6ff75bc8ce171c758bf75f304707
 * @param camelCaseString String in camelCase
 */
export const camelToKebab = (camelCaseString: string): string => {
  return camelCaseString
    .replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, '$1-$2')
    .toLowerCase();
};

/**
 * Build an href url by given template and properties,
 *
 * @example
 * const pathTemplate = '/resources/:selfUrl';
 * const selfUrl = '123'
 * const href = buildHref(pathTemplate, { selfUrl }) // /resources/123
 *
 * @param template [string]
 * @param props [object]
 */
export function buildHref(
  template: string,
  props: { [key: string]: string } = {},
) {
  return Object.keys(props).reduce((prev, key) => {
    const regExp = new RegExp(`:${key}`, 'g');
    return prev.replace(regExp, props[key]);
  }, template);
}
