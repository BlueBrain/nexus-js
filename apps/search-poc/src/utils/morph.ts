
import swcParser from 'swcmorphologyparser';


const CONVERTER_URL = 'http://morph-service.ocp.bbp.epfl.ch/converter/api';

export function convertToSwc(blob: Blob): Promise<String> {
  const formData = new FormData();
  formData.append('output_extension', '.swc');
  formData.append('isBlob', 'true');
  formData.append('file', new File([blob], 'file.asc'));

  return fetch(CONVERTER_URL, { method: 'POST', body: formData })
    .then(res => res.text())
}

export function parseSwc(swcMorph: string): Object {
  const parser = new swcParser.SwcParser()
  parser.parse(swcMorph);
  return parser.getRawMorphology();
}
