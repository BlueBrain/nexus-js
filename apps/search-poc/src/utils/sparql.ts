import { Filter } from '../components/Filters';
import { labelOf } from '../utils';
import get from 'lodash/get';

export interface Binding {
  [filterName: string]: { type: string; value: string };
}

export type SparqlQueryResults = {
  head: {
    vars: string[];
  };
  results: {
    bindings: Binding[];
  };
};

export type SparqlMapping = {
  source: string;
  target: string;
  defaultVal?: any;
};

export type SparqlMapperConfig = {
  mappings: SparqlMapping[];
};

// Transform Sparql results into a collection with properties from mapping config
// and optional default values.
export const mapSparqlResults = (
  queryResults: SparqlQueryResults,
  config: SparqlMapperConfig,
): { [target: string]: any }[] => {
  if (!queryResults || !queryResults.results.bindings) return [];

  return queryResults.results.bindings.map(binding => {
    const reduceFn = (acc, mapping) => ({
      ...acc,
      ...{
        [mapping.target]: get(
          binding,
          `${mapping.source}.value`,
          mapping.defaultVal,
        ),
      },
    });

    return config.mappings.reduce(reduceFn, {});
  });
};

// Method returns the collection with properties:
// { id: string, self: string, name: string, description: string }
export const mapEmodelCollQueryResults = (queryResults: SparqlQueryResults) => {
  const config = {
    mappings: [
      {
        source: 'id',
        target: 'emodel',
      },
      {
        source: 'self',
        target: 'self',
      },
      {
        source: 'name',
        target: 'name',
      },
      {
        source: 'description',
        target: 'description',
        defaultVal: 'NA',
      },
    ],
  };

  return mapSparqlResults(queryResults, config);
};

export const mapQueryResultsToFilters = (queryResults: SparqlQueryResults) => {
  return (
    queryResults &&
    queryResults.results &&
    queryResults.results.bindings.reduce((memo: Filter[], entry: Binding) => {
      const [filterIDData, filterLabelData] = Object.keys(entry);
      const filterName = filterIDData.replace('ID', '');
      const filter: Filter | undefined = memo.find(
        (entry: any) => entry.name === filterName,
      );
      const filterValue = {
        id: entry[filterIDData].value,
        label: filterLabelData
          ? entry[filterLabelData].value
          : labelOf(entry[filterIDData].value),
      };
      if (filter) {
        // don't push the same ID to the filter list
        // There may be repeats
        // and we don't want to display them
        if (
          filter.values.find(filter => filter.id === entry[filterIDData].value)
        ) {
          return memo;
        }

        // If it's not a repeat, add it to the values list
        filter.values.push(filterValue);

        // sort the values alphabetically (should be done last, but whatever)
        filter.values.sort((a, b) => {
          if (a.label < b.label) {
            return -1;
          }
          if (a.label > b.label) {
            return 1;
          }
          return 0;
        });
      } else {
        memo.push({
          name: filterName,
          values: [filterValue],
        });
      }
      return memo;
    }, [])
  );
};

export const makeDatasetFromSparlQueryResponse = (
  response: SparqlQueryResults,
) => {
  return response.results.bindings.reduce(
    (prev: { items: any[]; total: number }, binding: Binding) => {
      const [key] = Object.keys(binding);
      if (key === 'total') {
        prev.total = Number(binding[key].value);
      } else if (!binding[key]) {
        return prev;
      } else {
        const self = binding[key].value;
        if (self) {
          prev.items.push(self);
        }
      }
      return prev;
    },
    {
      items: [],
      total: 0,
    },
  );
};

export type DatasetQueryConfig = {
  vocab: string;
  graphs: { [filterName: string]: string };
};

export const makeDatasetQuery = (
  datasetQueryConfig: DatasetQueryConfig,
  appliedFilters: { [filterName: string]: string[] },
  size: number,
  from: number,
): string => {
  // graphQueries look like this
  // {
  //   brainRegion: `
  // Graph ?g {
  //   VALUES ?search { #{values} }
  //   ?s1 nxs:brainRegion ?search .
  // }
  //   `
  // }

  const graphQueries = Object.keys(datasetQueryConfig.graphs)
    .map(key => {
      // unparsed query looks like this
      // Graph ?g {
      //   VALUES ?search { #{values} }
      //   ?s1 nxs:brainRegion ?search .
      // }
      const unparsedQuery = datasetQueryConfig.graphs[key];

      // we dont need to parse this for now...
      // mainly because I don't know how to handle these query caes
      // where you always want it to be used
      if (key === 'constrainedBy') {
        return unparsedQuery;
      }

      // lets get the list of ids for this applied facet key
      // in our example, the key would be brainRegion
      // and the ids would be something like http://api.brain-map.org/api/v2/data/Structure/778
      const ids = appliedFilters[key];

      if (ids && ids.length) {
        // for the sparql query to work we need to always wrap URIs in <>
        const values = (ids || []).map(id => `<${id}>`).join(' ');
        const labels = (ids || []).map(id => `"${id}"`).join(' ');

        // now we need to add the appliedFacet values, if available
        const parseQuery = unparsedQuery
          .replace('#{values}', values)
          .replace('#{labels}', labels);
        return parseQuery;
      }
      return null;
    })
    .filter(queryString => !!queryString)
    .join('');

  return `
  ${datasetQueryConfig.vocab}

  SELECT ?total ?self
     WITH {
      SELECT DISTINCT ?self {
        ${graphQueries}
        Graph ?g {
          ?s nxv:self ?self
        }
      }
     } AS %resultSet

   WHERE {
        {
           SELECT (COUNT(?self) AS ?total)
           WHERE { INCLUDE %resultSet }
        }
        UNION
       {
           SELECT *
           WHERE { INCLUDE %resultSet }
           ORDER BY ?self
           LIMIT ${size}
           OFFSET ${from}
        }
     }
   `;
};
