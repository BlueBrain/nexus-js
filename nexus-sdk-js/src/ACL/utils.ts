import ACL from '.';
import { ACLResponse } from './types';
import { httpGet } from '../utils/http';

export async function getACL(
  path: string,
  rev?: number,
  self?: boolean,
): Promise<ACL> {
  try {
    const aclResponse: ACLResponse = await httpGet(`/acls/${path}`);
    return new ACL(aclResponse);
  } catch (error) {
    throw error;
  }
}
