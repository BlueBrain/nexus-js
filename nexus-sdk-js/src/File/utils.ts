import { FileResponse } from './types';
import { httpPost, HttpConfigTypes } from '../utils/http';
import { File as NexusFile } from '..';

export async function createFile(
  orgLabel: string,
  projectLabel: string,
  file: File,
): Promise<NexusFile> {
  try {
    const fileResponse: FileResponse = await httpPost<File>(
      `/files/${orgLabel}/${projectLabel}`,
      file,
      { sendAs: HttpConfigTypes.FILE },
    );
    return new NexusFile(orgLabel, projectLabel, fileResponse);
  } catch (error) {
    throw new Error(error);
  }
}
