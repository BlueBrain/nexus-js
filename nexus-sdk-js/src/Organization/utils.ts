import { Resource, Organization } from '..';
import { OrgResponse, ListOrgsResponse, ListOrgsOptions } from '.';
import { httpPut, httpGet, httpDelete } from '../utils/http';
import { CreateOrganizationException } from './exceptions';

/**
 *
 * @param label The label of your organization
 * @param name The name of your organization
 */
export async function createOrganization(
  label: string,
  name: string,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(`/orgs/${label}`, {
      name,
    });
    return new Organization({
      ...orgResponse,
      projectNumber: 0,
      _deprecated: false,
    });
  } catch (error) {
    throw new CreateOrganizationException(error.message);
  }
}

// TODO: refactor -> blocked by https://github.com/BlueBrain/nexus/issues/112
/**
 *
 * @param orgLabel The organization label to fetch
 * @param Object<revision, tag> The specific tag OR revision to fetch
 */
export async function getOrganization(
  orgLabel: string,
  options?: { revision?: number; tag?: string },
): Promise<Organization> {
  try {
    // check if we have options
    let ops = '';
    if (options) {
      // it's rev or tag, not both. We take rev over tag
      if (options.revision) {
        ops = `?rev=${options.revision}`;
      } else if (options.tag) {
        ops = `?tag=${options.tag}`;
      }
    }
    const orgResponse: OrgResponse = await httpGet(`/orgs/${orgLabel}${ops}`);
    // we want to know how many projects there are per organisation
    const listOrgsResponse: ListOrgsResponse = await httpGet('/projects');
    const projectNumber: number = listOrgsResponse._results
      ? listOrgsResponse._results.reduce((prev, org) => {
          const split = org._id.split('/');
          const orgName = split.slice(split.length - 2, split.length - 1)[0];
          if (orgName === orgLabel) {
            return prev + 1;
          }
          return prev;
        }, 0)
      : 0;

    const org = new Organization({ ...orgResponse, projectNumber });
    return org;
  } catch (e) {
    throw new Error(`ListOrgsError: ${e}`);
  }
}

// TODO: refactor -> blocked by https://github.com/BlueBrain/nexus/issues/112
// cannot implement list orgs options until then...
export async function listOrganizations(
  options?: ListOrgsOptions,
): Promise<Organization[]> {
  let ops;
  if (options) {
    ops = Object.keys(options).reduce(
      (currentOps, key) =>
        currentOps.length === 0
          ? `?${key}=${options[key]}`
          : `${currentOps}&${key}=${options[key]}`,
      '',
    );
  }
  try {
    const listOrgsResponse: ListOrgsResponse = await httpGet('/projects');
    if (listOrgsResponse.code || !listOrgsResponse._results) {
      return [];
    }
    // Get list of unique orgs names
    const filteredOrgNames: string[] = listOrgsResponse._results
      .map(org => {
        const split = org._id.split('/');
        const orgName = split.slice(split.length - 2, split.length - 1)[0];
        return orgName;
      })
      .filter((org, index, self) => self.indexOf(org) === index);

    // get orgs details
    return Promise.all(
      filteredOrgNames.map(async org => await getOrganization(org)),
    );
  } catch (e) {
    throw new Error(`ListOrgsError: ${e}`);
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last known revision
 * @param newName Knew name the organization will be called with
 */
export async function updateOrganization(
  orgLabel: string,
  rev: number = 1,
  newName: string,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}?rev=${rev}`,
      {
        name: newName,
      },
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 *
 * @param orgLabel Current organization label
 * @param rev Last know revision
 * @param Object<{tagName, revision}> The name of the tag and revision number to tag the organization from
 */
export async function tagOrganization(
  orgLabel: string,
  rev: number = 1,
  {
    tagName,
    tagFromRev,
  }: {
    tagName: string;
    tagFromRev: number;
  },
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpPut(
      `/orgs/${orgLabel}/tags?rev=${rev}`,
      {
        tag: tagName,
        rev: tagFromRev,
      },
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 *
 * @param orgLabel Label of the Org to be deprecated
 * @param rev Last know revision
 */
export async function deprecateOrganization(
  orgLabel: string,
  rev: number = 1,
): Promise<Organization> {
  try {
    const orgResponse: OrgResponse = await httpDelete(
      `/orgs/${orgLabel}?rev=${rev}`,
    );
    return new Organization(orgResponse);
  } catch (error) {
    throw new Error(error);
  }
}
