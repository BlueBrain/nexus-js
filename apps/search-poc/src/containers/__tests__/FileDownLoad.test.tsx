import FileDownload from '../FileDownLoad';
import React from "react";
import {  mount } from "enzyme";
import { BrowserRouter as Router } from 'react-router-dom';
import { NexusProvider } from '@bbp/react-nexus';
import { NexusClient, createNexusClient } from '@bbp/nexus-sdk';
import { setToken } from '../../utils/auth';

describe('FileDownload', () => {
    
    it('FileDownload renders correctly', () => {
        const nexus: NexusClient = createNexusClient({
            fetch,
            uri: 'test',
            links: [setToken],
          });
        const mockCallBack = jest.fn();
        const props = {
            fileIds: ['path/org/project/test-id'],
            isDownload: true,
            downloadFileName: 'test-download',
            OnShowFiles: mockCallBack
          };
        const result = mount(<NexusProvider nexusClient={nexus}><Router><FileDownload {...props} /></Router></NexusProvider>)
        expect(result).toMatchSnapshot();
    });

    it('OnShowFiles is called when toggle  is clicked', () => {
        const nexus: NexusClient = createNexusClient({
            fetch,
            uri: 'test',
            links: [setToken],
          });
        
        const mockCallBack = jest.fn();  
        const props = {
            fileIds: ['path/org/project/test-id'],
            isDownload: true,
            downloadFileName: 'test-download',
            OnShowFiles: mockCallBack
          };
        const result = mount(<NexusProvider nexusClient={nexus}><Router><FileDownload {...props} /></Router></NexusProvider>)
        result.find('button').at(0).simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('the download icon changes to loading when download button is clicked', () => {
        const nexus: NexusClient = createNexusClient({
            fetch,
            uri: 'test',
            links: [setToken],
          });
        
        const mockCallBack = jest.fn();  
        const props = {
            fileIds: ['path/org/project/test-id'],
            isDownload: true,
            downloadFileName: 'test-download',
            OnShowFiles: mockCallBack
          };
        const result = mount(<NexusProvider nexusClient={nexus}><Router><FileDownload {...props} /></Router></NexusProvider>)
        result.find('button').at(1).simulate('click');
        expect(result).toMatchSnapshot();
    });

});

