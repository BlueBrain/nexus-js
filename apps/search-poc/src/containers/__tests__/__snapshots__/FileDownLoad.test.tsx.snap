// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileDownload FileDownload renders correctly 1`] = `
<NexusProvider
  nexusClient={
    Object {
      "ACL": Object {
        "append": [Function],
        "create": [Function],
        "delete": [Function],
        "list": [Function],
        "poll": [Function],
        "replace": [Function],
        "subtract": [Function],
      },
      "Archive": Object {
        "create": [Function],
        "createWithId": [Function],
        "get": [Function],
      },
      "File": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "link": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Identity": Object {
        "list": [Function],
        "poll": [Function],
      },
      "Organization": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Permissions": Object {
        "append": [Function],
        "delete": [Function],
        "list": [Function],
        "poll": [Function],
        "replace": [Function],
        "subtract": [Function],
      },
      "Project": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Realm": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Resolver": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Resource": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "links": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Schema": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Storage": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "View": Object {
        "create": [Function],
        "deprecate": [Function],
        "elasticSearchQuery": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "pollStatistics": [Function],
        "sparqlQuery": [Function],
        "statistics": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "httpDelete": [Function],
      "httpGet": [Function],
      "httpPatch": [Function],
      "httpPost": [Function],
      "httpPut": [Function],
      "poll": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <FileDownloadContainer
        OnShowFiles={[MockFunction]}
        downloadFileName="test-download"
        fileIds={
          Array [
            "path/org/project/test-id",
          ]
        }
        isDownload={true}
      >
        <div
          className="file-download"
        >
          <div
            className="toggle"
          >
            <Switch
              checked={true}
              onChange={[Function]}
              size="small"
            >
              <Wave
                insertExtraNode={true}
              >
                <Switch
                  checked={true}
                  checkedChildren={null}
                  className="ant-switch-small"
                  defaultChecked={false}
                  loadingIcon={null}
                  onChange={[Function]}
                  prefixCls="ant-switch"
                  size="small"
                  unCheckedChildren={null}
                >
                  <button
                    aria-checked={true}
                    checked={true}
                    className="ant-switch-small ant-switch ant-switch-checked"
                    defaultChecked={false}
                    onChange={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onMouseUp={[Function]}
                    role="switch"
                    size="small"
                    type="button"
                  >
                    <span
                      className="ant-switch-inner"
                    />
                  </button>
                </Switch>
              </Wave>
            </Switch>
            <label>
              Files
            </label>
          </div>
          <div
            className="download"
          >
            <a
              download="test-download.tar.gz"
              href={null}
              style={
                Object {
                  "display": "none",
                }
              }
            >
              download
            </a>
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              icon="download"
              loading={false}
              onBlur={[Function]}
              onClick={[Function]}
              size="small"
            >
              <Wave>
                <button
                  className="ant-btn ant-btn-sm"
                  onBlur={[Function]}
                  onClick={[Function]}
                  type="button"
                >
                  <Icon
                    type="download"
                  >
                    <LocaleReceiver
                      componentName="Icon"
                    >
                      <i
                        aria-label="icon: download"
                        className="anticon anticon-download"
                      >
                        <IconReact
                          className=""
                          type="download-o"
                        >
                          <svg
                            aria-hidden="true"
                            className=""
                            data-icon="download"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            key="svg-download"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M505.7 661a8 8 0 0 0 12.6 0l112-141.7c4.1-5.2.4-12.9-6.3-12.9h-74.1V168c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v338.3H400c-6.7 0-10.4 7.7-6.3 12.9l112 141.8zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                              key="svg-download-svg-0"
                            />
                          </svg>
                        </IconReact>
                      </i>
                    </LocaleReceiver>
                  </Icon>
                  <span
                    key=".0"
                  >
                    Download 1 file(s)
                  </span>
                </button>
              </Wave>
            </Button>
          </div>
        </div>
      </FileDownloadContainer>
    </Router>
  </BrowserRouter>
</NexusProvider>
`;

exports[`FileDownload the download icon changes to loading when download button is clicked 1`] = `
<NexusProvider
  nexusClient={
    Object {
      "ACL": Object {
        "append": [Function],
        "create": [Function],
        "delete": [Function],
        "list": [Function],
        "poll": [Function],
        "replace": [Function],
        "subtract": [Function],
      },
      "Archive": Object {
        "create": [Function],
        "createWithId": [Function],
        "get": [Function],
      },
      "File": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "link": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Identity": Object {
        "list": [Function],
        "poll": [Function],
      },
      "Organization": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Permissions": Object {
        "append": [Function],
        "delete": [Function],
        "list": [Function],
        "poll": [Function],
        "replace": [Function],
        "subtract": [Function],
      },
      "Project": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Realm": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "update": [Function],
      },
      "Resolver": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Resource": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "links": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Schema": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "Storage": Object {
        "create": [Function],
        "deprecate": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "View": Object {
        "create": [Function],
        "deprecate": [Function],
        "elasticSearchQuery": [Function],
        "get": [Function],
        "list": [Function],
        "poll": [Function],
        "pollStatistics": [Function],
        "sparqlQuery": [Function],
        "statistics": [Function],
        "tag": [Function],
        "update": [Function],
      },
      "httpDelete": [Function],
      "httpGet": [Function],
      "httpPatch": [Function],
      "httpPost": [Function],
      "httpPut": [Function],
      "poll": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <FileDownloadContainer
        OnShowFiles={[MockFunction]}
        downloadFileName="test-download"
        fileIds={
          Array [
            "path/org/project/test-id",
          ]
        }
        isDownload={true}
      >
        <div
          className="file-download"
        >
          <div
            className="toggle"
          >
            <Switch
              checked={true}
              onChange={[Function]}
              size="small"
            >
              <Wave
                insertExtraNode={true}
              >
                <Switch
                  checked={true}
                  checkedChildren={null}
                  className="ant-switch-small"
                  defaultChecked={false}
                  loadingIcon={null}
                  onChange={[Function]}
                  prefixCls="ant-switch"
                  size="small"
                  unCheckedChildren={null}
                >
                  <button
                    aria-checked={true}
                    checked={true}
                    className="ant-switch-small ant-switch ant-switch-checked"
                    defaultChecked={false}
                    onChange={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onMouseUp={[Function]}
                    role="switch"
                    size="small"
                    type="button"
                  >
                    <span
                      className="ant-switch-inner"
                    />
                  </button>
                </Switch>
              </Wave>
            </Switch>
            <label>
              Files
            </label>
          </div>
          <div
            className="download"
          >
            <a
              download="test-download.tar.gz"
              href={null}
              style={
                Object {
                  "display": "none",
                }
              }
            >
              download
            </a>
            <Button
              block={false}
              ghost={false}
              htmlType="button"
              icon="loading"
              loading={true}
              onBlur={[Function]}
              onClick={[Function]}
              size="small"
            >
              <Wave>
                <button
                  className="ant-btn ant-btn-sm ant-btn-loading"
                  onBlur={[Function]}
                  onClick={[Function]}
                  type="button"
                >
                  <Icon
                    type="loading"
                  >
                    <LocaleReceiver
                      componentName="Icon"
                    >
                      <i
                        aria-label="icon: loading"
                        className="anticon anticon-loading"
                      >
                        <IconReact
                          className="anticon-spin"
                          type="loading-o"
                        >
                          <svg
                            aria-hidden="true"
                            className="anticon-spin"
                            data-icon="loading"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            key="svg-loading"
                            viewBox="0 0 1024 1024"
                            width="1em"
                          >
                            <path
                              d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 0 0-94.3-139.9 437.71 437.71 0 0 0-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                              key="svg-loading-svg-0"
                            />
                          </svg>
                        </IconReact>
                      </i>
                    </LocaleReceiver>
                  </Icon>
                  <span
                    key=".0"
                  >
                    Download 1 file(s)
                  </span>
                </button>
              </Wave>
            </Button>
          </div>
        </div>
      </FileDownloadContainer>
    </Router>
  </BrowserRouter>
</NexusProvider>
`;
